akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala

```diff
@@ -324,10 +326,19 @@ private[client] object NewHostConnectionPool {
 
           def settings: ConnectionPoolSettings = _settings
 
+          private lazy val keepAliveDurationFuzziness: () ⇒ Long = {
+            val random = new Random()
+            val max = (settings.maxConnectionKeepAliveTime / 10).toMillis
+            () ⇒ random.nextLong() % max
+          }
           def openConnection(): Future[Http.OutgoingConnection] = {
             if (connection ne null) throw new IllegalStateException("Cannot open connection when slot still has an open connection")
 
             connection = logic.openConnection(this)
+            connection.outgoingConnection.foreach(_ ⇒ if (settings.maxConnectionKeepAliveTime.isFinite()) {
+
+              disconnectAt = Instant.now().getEpochSecond + settings.maxConnectionKeepAliveTime.toMillis + keepAliveDurationFuzziness()
```

Comment:
Hmm is this thread-safe? `connection.outgoingConnection.foreach` might not be strictly needed here and make things simpler?

Reply:
I guess you are right, that `foreach` does not do much